lib/Runtime/ByteCode/ByteCodeEmitter.cpp

```diff
@@ -10302,7 +10334,15 @@ void EmitYieldStar(ParseNodeUni* yieldStarNode, ByteCodeGenerator* byteCodeGener
     funcInfo->ReleaseLoc(yieldStarNode->pnode1);
 
     // Call the iterator's next()
-    EmitIteratorNext(yieldStarNode->location, iteratorLocation, funcInfo->undefinedConstantRegister, byteCodeGenerator, funcInfo);
+    CacheIteratorNext(iteratorLocation, nextCache, byteCodeGenerator, funcInfo);
+
+    byteCodeGenerator->Writer()->Reg1(Js::OpCode::NewScObjectSimple, iteratorObject);
+    uint cacheId = funcInfo->FindOrAddInlineCacheId(iteratorObject, Js::PropertyIds::_symbolIterator, false, true);
+    byteCodeGenerator->Writer()->PatchableProperty(Js::OpCode::StFld, iteratorLocation, iteratorObject, cacheId);
+    cacheId = funcInfo->FindOrAddInlineCacheId(iteratorObject, Js::PropertyIds::next, false, true);
+    byteCodeGenerator->Writer()->PatchableProperty(Js::OpCode::StFld, nextCache, iteratorObject, cacheId);
```

Comment:
If I understand correctly, this will create a temporary object whose `Symbol.iterator` property will be the "real" iterator object and whose `next` property is the cached next method. And this object is ultimately used in `JavascriptOperators::OP_ResumeYield`.

A couple of thoughts:

- Symbol.iterator would need to be a method (if it's used at all by `OP_ResumeYield`).
- Will calling the next method from `OP_ResumeYield` result in the correct this value? Or will the this value be the wrapper iterator?
- Does `OP_ResumeYield` also expect "throw" and "return" to be present on the wrapper iterator (if they exist on the original iterator)?

Let's think about how we would want this to work ideally. Perhaps `OP_ResumeYield` should expect an object of the following form instead?

```
{
  iterator: Var; // The yield star iterator
  next: JavascriptFunction* // The cached next method
}
```

@boingoing Thoughts?

Reply:
It uses symbol.iterator as a convenient label the object it puts there is the yield* iterator and not the orriginal Symbol.iterator

Over in ResumeYield it is retrieved as iteratorObject and correctly used as the this value, it is not called.